### PR TITLE
File to execute AFNI 3dautomask algorithm

### PR DESCRIPTION
**Pull request by Lestropie**
Constructing a draft PR to show the progress being made by @RicardoRios46.

- [x] Name: Should algorithm name be `3dautomask` or `afni`? Are there other commands within AFNI that would plausibly be used for brain mask derivation and therefore the more specific name is necessary?

- [x] Fix algorithm author

- [x] Add any relevant citations

- [x] Add command-line options to influence the AFNI command (or, if there's a long list of such options, just give the user the ability to provide a string at the command-line that will be passed down to the AFNI command as-is)

### CODE REVIEW

File: lib/mrtrix3/dwi2mask/3dautomask.py [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+
+from distutils.spawn import find_executable
+from mrtrix3 import MRtrixError
+from mrtrix3 import app, image, path, run
+
+
+
+def usage(base_parser, subparsers): #pylint: disable=unused-variable
+  parser = subparsers.add_parser('3dautomask', parents=[base_parser])
+  parser.set_author('Robert E. Smith (robert.smith@florey.edu.au) and Ricardo Rios (ricardo.rios@cimat.mx)')
```
Comment (Lestropie) [inline lib/mrtrix3/dwi2mask/3dautomask.py:24]: Happy to be left off this one entirely and for you to be the sole author.

File: lib/mrtrix3/dwi2mask/3dautomask.py [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+  options.add_argument('-SI', type=float, help='After creating the mask, find the most superior voxel, then zero out everything more than SI millimeters inferior to that. 130 seems to be decent (i.e., for Homo Sapiens brains).')
+                 
+  #ToDo Are the next parameters actually floats?
+  options.add_argument('-dilate', type=float, help='Dilate the mask outwards n times')
+  options.add_argument('-erode ', type=float, help='Erode  the mask outwards n times')
+  
+  # ToDo. Maybe there is a better way to code these related inputs (they are not mutually exclusive tough)
+  options.add_argument('-NN1', action='store_true', help='Erode and dilate using different neighbor definitions NN1=faces, NN2=edges, NN3= corners')   
+  options.add_argument('-NN2', action='store_true', help='Erode and dilate using different neighbor definitions NN1=faces, NN2=edges, NN3= corners')   
+  options.add_argument('-NN3', action='store_true', help='Erode and dilate using different neighbor definitions NN1=faces, NN2=edges, NN3= corners')
```
Comment (Lestropie) [inline lib/mrtrix3/dwi2mask/3dautomask.py:44]: The help page kinda looks like these are mutually exclusive, but it's unclear.

If that is in fact the case, you can use `parser.flag_mutually_exclusive_options()`.

File: lib/mrtrix3/dwi2mask/3dautomask.py [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+  options.add_argument('-nograd', action='store_true', help='The program uses a \'gradual\' clip level by default. Add this option to use a fixed clip level.')
+  options.add_argument('-peels', type=float, help='Peel (erode) the mask n times, then unpeel (dilate).')
+  options.add_argument('-nbhrs', type=int, help='Define the number of neighbors needed for a voxel NOT to be eroded.  It should be between 6 and 26.')
+  options.add_argument('-eclip', action='store_true', help='After creating the mask, remove exterior voxels below the clip threshold.')   
+  options.add_argument('-SI', type=float, help='After creating the mask, find the most superior voxel, then zero out everything more than SI millimeters inferior to that. 130 seems to be decent (i.e., for Homo Sapiens brains).')
+                 
+  #ToDo Are the next parameters actually floats?
+  options.add_argument('-dilate', type=float, help='Dilate the mask outwards n times')
+  options.add_argument('-erode ', type=float, help='Erode  the mask outwards n times')
```
Comment (Lestropie) [inline lib/mrtrix3/dwi2mask/3dautomask.py:39]: Looks to me like they should be `int`s

File: lib/mrtrix3/dwi2mask/3dautomask.py [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+  if app.ARGS.nograd:
+    cmd_string += ' -nograd '
+  if app.ARGS.eclip:
+    cmd_string += ' -eclip '
+  if app.ARGS.NN1:
+    cmd_string += ' -NN1 ' 
+  if app.ARGS.NN2:
+    cmd_string += ' -NN2 '  
+  if app.ARGS.NN3:
+    cmd_string += ' -NN3 '
```
Comment (Lestropie) [inline lib/mrtrix3/dwi2mask/3dautomask.py:81]: These will end up with excess spaces in the command string; if you stick to the convention of *never* leaving an extra space at the end of the string, and *always* inserting a space prior to any append operations, you should get the correct result with no double-spaces regardless of usage.

File: lib/mrtrix3/dwi2mask/3dautomask.py (hunk 2 above)
Comment (RicardoRios46) [reply lib/mrtrix3/dwi2mask/3dautomask.py:44]: yeah I will  run AFNI with all these flags and see if trows an error or something

File: lib/mrtrix3/dwi2mask/3dautomask.py (hunk 4 above)
Comment (RicardoRios46) [reply lib/mrtrix3/dwi2mask/3dautomask.py:81]: Ok, if the  str(app.ARGS.) does that then all good.

File: lib/mrtrix3/dwi2mask/3dautomask.py (hunk 3 above)
Comment (RicardoRios46) [reply lib/mrtrix3/dwi2mask/3dautomask.py:39]: That's what I though but no errors or warnings if I use floats. I'll try to see if I can turn on more verbatim in the execution or something

Comment (Lestropie) [reply lib/mrtrix3/dwi2mask/3dautomask.py:39]: It probably gets immediately cast to `int` in their own command-line parsing code. However in *MRtrix3*, for options that receive a compulsory argument (e.g. "`-dilate 2`"), the data type by which that information will be interpreted is shown in the help page alongside the name of the option. So I think it makes sense to expect an `int` for these options and also indicate to users that that is the type of data expected, even if passing a `float` to AFNI technically doesn't make it crash.

File: lib/mrtrix3/dwi2mask/3dautomask.py (hunk 1 above)
Comment (Lestropie) [reply lib/mrtrix3/dwi2mask/3dautomask.py]: Addressed in ff739133.

File: lib/mrtrix3/dwi2mask/3dautomask.py (hunk 3 above)
Comment (Lestropie) [reply lib/mrtrix3/dwi2mask/3dautomask.py]: Resolved in ff739133.

File: lib/mrtrix3/dwi2mask/3dautomask.py (hunk 4 above)
Comment (Lestropie) [reply lib/mrtrix3/dwi2mask/3dautomask.py]: Resolved in ff739133.

**Comment by Lestropie** [comment]
While I've not had success installing AFNI on my exotic Linux system, the code looks correct, has been resolved against other parallel changes, and passes `pylint`. So I'm going to merge this algorithm into the `dwi2mask` branch ahead of the completion of the hackathon. Subsequent changes or fixes are of course possible.

**Pull request merged**